core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -1,187 +1,337 @@
+                    let mut wg = map.write().unwrap();
+                    if wg.capacity() > validator_stakes.len() * 2 {
+                        // map is too large for what we need currently, reallocate it
+                        *wg = map_maker();
+                    } else {
+                        // reinit the existing memory with up-to-date validator set
+                        wg.clear();
+                        zero_init_map(&mut wg);
+                    }
+                    drop(wg);
+                    map
+                }
+                // if we can not reuse just allocate a new one =(
+                Err(_) => RwLock::new(map_maker()),
+            };
+            cluster_slots.push_back(RowContent {
+                slot,
+                total_support: AtomicU64::new(0),
+                supporters: Arc::new(map),
+            });
         }
-        self.report_cluster_slots_size();
+        debug_assert!(
+            cluster_slots.len() == CLUSTER_SLOTS_TRIM_SIZE,
+            "Ring buffer should be exactly the intended size"
+        );
     }
 
-    fn report_cluster_slots_size(&self) {
-        if self.last_report.should_update(10_000) {
-            let (cluster_slots_cap, pubkeys_capacity) = {
-                let cluster_slots = self.cluster_slots.read().unwrap();
-                let cluster_slots_cap = cluster_slots.len();
-                let pubkeys_capacity = cluster_slots
-                    .iter()
-                    .map(|(_slot, slot_pubkeys)| slot_pubkeys.read().unwrap().capacity())
-                    .sum::<usize>();
-                (cluster_slots_cap, pubkeys_capacity)
-            };
-            let (validator_stakes_cap, validator_pubkeys_len) = {
-                let validator_stakes = self.validator_stakes.read().unwrap();
-                let validator_len = validator_stakes
-                    .iter()
-                    .map(|(_pubkey, vote_accounts)| vote_accounts.vote_accounts.capacity())
-                    .sum::<usize>();
-                (validator_stakes.capacity(), validator_len)
+    fn update_internal(
+        &self,
+        root: Slot,
+        validator_stakes: &HashMap<Pubkey, u64>,
+        epoch_slots_list: Vec<EpochSlots>,
+    ) {
+        let total_stake = self.total_stake.load(std::sync::atomic::Ordering::Relaxed);
+        // Prepare a range of slots we will store in the datastructure
+        // We will discard slots at or before current root or too far ahead.
+        let slot_range = (root + 1)..root.saturating_add(CLUSTER_SLOTS_TRIM_SIZE as u64 + 1);
+        // ensure the datastructure has the correct window in scope
+        self.roll_cluster_slots(validator_stakes, &slot_range);
+
+        let cluster_slots = self.cluster_slots.read().unwrap();
+        for epoch_slots in epoch_slots_list {
+            //filter out unstaked nodes
+            let Some(&sender_stake) = validator_stakes.get(&epoch_slots.from) else {
+                continue;
```

Comment:
Now we are explicitly ignoring epoch slots from unstaked nodes here to limit the datastructure size